### ISSUE TITLE
Showing almost Zero Accuracy During Inference

### ISSUE THREAD

**Issue by joymahapatra** [itiv-kit/cnn-parted#1]
Hello, 
first of all, thank you for sharing this fantastic repository! I am currently trying to run the code and wanted to monitor accuracy during inference. For this, I made some monitoring modifications (commented with "added this") to the `tools/dnn-model-exploration/model_explorer/accuracy_functions/classification_accuracy.py` file as follows:

```python
    correct_pred = 0
    total_prediction = 0 # added this
    model.eval()
    with torch.no_grad():
        for X, y_true in dataloader:
            X = X.to(device)
            y_true = y_true.to(device)

            y_prob = model(X)
            _, predicted_labels = torch.max(y_prob, 1)
            correct_pred += (predicted_labels == y_true).sum()
            total_prediction += torch.numel(predicted_labels) # added this
            progress_bar.update(y_true.size(0))

    correct_pred = correct_pred.to(cpu_device)
    print(f"Total number of correct predictions: {correct_pred}") # added this
    print(f"Total number of predictions: {total_prediction}") # added this
    return correct_pred.float() / dataset_size
```

With this code, I noticed that no accuracy is being displayed during inference (see the attached screenshot). I’m unable to understand why a pre-trained vision model isn’t showing any accuracy. I tested this with both SqueezeNet and ResNet18, but the issue persists.

![Screenshot from 2024-12-05 00-38-18](https://github.com/user-attachments/assets/708e1be6-3a58-4ae5-81db-d5ff96e73e26)

Is there an alternative approach to check the model’s accuracy? I would appreciate any guidance or suggestions to resolve this issue.

**Comment by patschmidt2** [comment]
Hi, 
could you expand a bit on the test case you are trying to run, i.e. the YAML config? I have tested the accuracy evaluation for a ResNet18 and that seems to work on my side. The actual accuracy of the model is written to the CSV files that contain all the results. These files are generated at the end of the run, but it will take a few hours until the tool has completed the evaluation. For the default ResNet18 run that is found in the `examples` folder, it took about 6 hours to complete for me.

**Comment by joymahapatra** [comment]
Hello @patschmidt2 , Ive considered the following configuration for ResNet18 (`resnet18.yaml`) on a small subset of the ImageNet dataset.

```yaml
general:
  fixed_sys: False
  acc_once: False
  num_pp: -1
  num_topos: 1
  num_jobs: 1
  work_dir: wd_resnet
  optimization: ppa

workload:
  name: resnet18
  input-size: [ 1, 3, 227, 227 ]

pipeline:
  stages:
    - WorkloadParser
    - SystemParser
    - GraphAnalysis
    - RobustnessOptimization
    - NodeEvaluation
    - PartitioningOptimization
    - AccuracyEvaluation
    - ExportPartitionResults

accuracy:
  robustness:
    bits: [4, 6, 8, 16]
    delta : 0.01 # best_acc - delta = lower boundary for the selection of the bit constraints
  retraining:
    epochs: 0
  datasets:
    calibrate:
      type: imagenet
      kind: 'imagefolder'
      path: '/datasets/sub_imagenet/val'
      sample_limit: 50
      batch_size: 32
      randomize: False
      file: './calibrations/resnet18_quant_calib.pkl'
    validation:
      type: imagenet
      kind: 'imagefolder'
      path: '/datasets/sub_imagenet/val'
      sample_limit: 50 # number of validation dataset samples to test for each individual
      batch_size: 32
      randomize: False
    train:
      type: imagenet
      kind: 'imagefolder'
      path: '/datasets/sub_imagenet/train'
      total_samples: 200000
      batch_size: 256
      sample_limit: null

components:
  - id: 0
    bits: 16
    fault_rates: [0, 0] # input, weight
    max_memory_size: 600000000
    timeloop:
      accelerator: eyeriss_like
      frequency: 200000000
      layer: conv2d
      mapper:
        algorithm: random-pruned
        live-status: false
        max-permutations-per-if-visit: 16
        num-threads: 8
        optimization-metrics:
        - edp
        timeout: 15000
        victory-condition: 500
  - id: 1
    max_out_size: 20000000
    data_bit_width: 8
    noi:
      noi_mode: UCIe
      width: 16
      data_rate_Gbps: 4
      latency_ns: 2 # including D2D Adapter and PHY
      power_bit_pj: 0.5
    fps: 25
  - id: 2
    bits: 8
    max_memory_size: 600000000
    timeloop:
      accelerator: simba_like
      frequency: 200000000
      layer: conv2d
      mapper:
        algorithm: random-pruned
        live-status: false
        max-permutations-per-if-visit: 16
        num-threads: 8
        optimization-metrics:
        - edp
        timeout: 15000
        victory-condition: 500
  - id: 3
    fault_rates: [0, 1e-4] # input, weight
    max_memory_size: 600000000
    mnsim:
      accelerator: MRAM_area # accelerator name
      conf_path: '/configs/mnsim_configs/MRAM_area.ini'
  - id: 4
    fault_rates: [0, 1e-4] # input, weight
    max_memory_size: 600000000
    mnsim:
      accelerator: MRAM_lat # accelerator name
      conf_path: '/configs/mnsim_configs/MRAM_lat.ini'
  - id: 5
    fault_rates: [0, 1e-4] # input, weight
    max_memory_size: 600000000
    mnsim:
      accelerator: RRAM_acc # accelerator name
      conf_path: '/configs/mnsim_configs/RRAM_acc.ini'
```

However, I’m still struggling to understand why the accuracy isn’t showing at all from the beginning, despite using a **pre-trained** ResNet18 model. If you’re indeed using a pre-trained ResNet18 for inference, it should display accuracy right from initialization. Am I misunderstanding something?

**Comment by patschmidt2** [comment]
The `RobustnessOptimization` stage tries to minimize the bit width of the layers as far as possible, so the accuracy will change compared to the pre-trained model.
